sklearn/inspection/_plot/partial_dependence.py

```diff
@@ -673,55 +722,69 @@ def plot(self, ax=None, n_cols=3, line_kw=None, contour_kw=None):
-                            values[0], ins.ravel(), **individual_line_kw
+                    ice_lines_idx = rng.choice(
+                        ice_lines.shape[0], n_ice_to_plot, replace=False
+                    )
+                    ice_lines_subsampled = ice_lines[ice_lines_idx, :]
+                    for ice_idx, ice in enumerate(ice_lines_subsampled):
+                        line_idx = pd_plot_idx * n_lines + ice_idx
+                        lines_ravel[line_idx] = axi.plot(
+                            values[0], ice.ravel(), **individual_line_kw
                         )[0]
-                if self.kind == 'average':
-                    lines_ravel[i] = axi.plot(
-                        values[0], avg_preds[self.target_idx].ravel(),
-                        **line_kw
-                    )[0]
-                elif self.kind == 'both':
-                    lines_ravel[i] = axi.plot(
-                        values[0], avg_preds[self.target_idx].ravel(),
-                        label='average', **line_kw
+                if self.kind in ("average", "both"):
+                    label = None if self.kind == "average" else "average"
+                    # the average is stored as the last line
+                    pd_line_idx = (
+                        pd_plot_idx
+                        if self.kind == "average"
+                        else pd_plot_idx * n_lines + n_ice_to_plot
+                    )
+                    lines_ravel[pd_line_idx] = axi.plot(
+                        values[0],
+                        avg_preds[self.target_idx].ravel(),
+                        label=label,
+                        **line_kw,
                     )[0]
-                    axi.legend()
             else:
                 # contour plot
                 XX, YY = np.meshgrid(values[0], values[1])
                 Z = avg_preds[self.target_idx].T
-                CS = axi.contour(XX, YY, Z, levels=Z_level, linewidths=0.5,
-                                 colors='k')
-                contours_ravel[i] = axi.contourf(XX, YY, Z, levels=Z_level,
-                                                 vmax=Z_level[-1],
-                                                 vmin=Z_level[0],
-                                                 **contour_kw)
-                axi.clabel(CS, fmt='%2.2f', colors='k', fontsize=10,
-                           inline=True)
+                CS = axi.contour(
+                    XX, YY, Z, levels=Z_level, linewidths=0.5, colors="k"
+                )
+                contours_ravel[pd_plot_idx] = axi.contourf(
+                    XX,
+                    YY,
+                    Z,
+                    levels=Z_level,
+                    vmax=Z_level[-1],
+                    vmin=Z_level[0],
+                    **contour_kw,
+                )
+                axi.clabel(
+                    CS, fmt="%2.2f", colors="k", fontsize=10, inline=True
+                )
```

Comment:
I think all that is related to the `len(values) != 1` case should be extracted in a dedicated `_plot_pairwise_dependence_contour` to improve readability (even if that might involve a bit of duplication for the labels of the axis).